fix: Throw UnsupportedTypeError from CompositePayloadConverter

lorensr opened the pull request:
Previously, `JsonPayloadConverter` throws, which prevents it from being anything but last in a `CompositePayloadConverter` order. `CompositePayloadConverter` also returned `undefined` when unable to convert. 

Now,  `JsonPayloadConverter` returns `undefined`, so it can be used in any order, and `CompositePayloadConverter` throws `UnsupportedTypeError` when unable to convert.

bergundy commented on packages/common/src/converter/payload-converters.ts:
```diff
@@ -1,11 +1,65 @@
+  readonly converterByEncoding: Map<string, PayloadConverterWithEncoding> = new Map();
+
+  constructor(...converters: PayloadConverterWithEncoding[]) {
+    this.converters = converters;
```
Let’s verify that this in a non empty array

bergundy commented on packages/internal-workflow-common/src/errors.ts:
```diff
@@ -6,12 +6,8 @@ export class PayloadConverterError extends ValueError {
-  constructor(message: string | undefined, public readonly cause?: Error) {
-    super(message ?? undefined);
-  }
+export class UnsupportedTypeError extends PayloadConverterError {
```
Why not reuse TypeError?

lorensr replied:
Generally, I'd make more specific errors in case a dev would want to do `instanceof` to differentiate. In this case, I don't imagine that being common—would be useful if the dev doesn't know what type of arg they're sending (eg it's provided by the user or something).

bergundy reviewed (approved):
Some minor comments.
I like what you did with the json error

lorensr commented:
- Changed `PayloadConverter` interface:

```
-  public abstract toPayload<T>(value: T): Payload | undefined;
+  public abstract toPayload<T>(value: T): Payload;
```

So that implementors don't think that returning `undefined` is valid. 

- Further, added check that custom `PayloadConverter`s toPayload returns a Payload-like object (containing either a `metadata` or `data` field).

lorensr commented on packages/workflow/src/worker-interface.ts:
```diff
@@ -128,7 +129,7 @@ export async function initRuntime({ info, randomnessSeed, now, patches }: Workfl
   // The `payloadConverter` export is validated in the Worker
   if (customPayloadConverter !== undefined) {
-    state.payloadConverter = customPayloadConverter;
+    state.payloadConverter = new WrappedCustomPayloadConverter(customPayloadConverter);
```
I missed this the first go around. Eventually found it through the `toPayload throwing results in task retry` test

lorensr commented:
Per further discussion:
- Changed signature back to `Payload | undefined`
- Removed Wrapped class, instead:
  - throw from toPayload helper
  - searched for all instances of `.toPayload` and switched non-test instances to use the helper

bergundy commented on packages/common/src/converter/payload-converters.ts:
```diff
@@ -87,3 +141,27 @@ export class BinaryPayloadConverter implements PayloadConverterWithEncoding {
+ *
+ * `const myConverter = new DefaultPayloadConverter({ protobufRoot })`
+ */
+export const defaultPayloadConverter = new WrappedPayloadConverter(new DefaultPayloadConverter());
```
Let's not wrap the default, instead I'd wrap in the worker / workflow runtime / client, that way any converter we accept would be wrapped.

lorensr replied:
Like this?

https://github.com/temporalio/sdk-typescript/pull/558/commits/a59fe705c366852c043ffb4dcba6338c2847187a

bergundy commented on packages/common/src/converter/payload-converters.ts:
```diff
@@ -1,19 +1,75 @@
-import { errorMessage, UnsupportedJsonTypeError, ValueError } from '@temporalio/internal-workflow-common';
```
All of the helper methods here should accept WrappedPayloadConverter to make sure we never use an unwrapped one IMHO.
WDYT?

bergundy reviewed (commented):
Skimmed this, overall LGTM, had a couple comments.

bergundy commented on packages/common/src/converter/payload-converters.ts:
```diff
@@ -87,3 +141,27 @@ export class BinaryPayloadConverter implements PayloadConverterWithEncoding {
+  // {@link DefaultPayloadConverterWithProtobufs}.
+  //
+  // Go SDK:
+  // https://github.com/temporalio/sdk-go/blob/5e5645f0c550dcf717c095ae32c76a7087d2e985/converter/default_data_converter.go#L28
```
Do we need this link?

lorensr replied:
I like it as a reference to what the order should be. Until we have a spec 😄

bergundy commented on packages/internal-non-workflow-common/src/otel.ts:
```diff
@@ -41,7 +44,7 @@ export async function extractSpanContextFromHeaders(headers: Headers): Promise<o
   const carrier = {};
   otel.propagation.inject(otel.context.active(), carrier, otel.defaultTextMapSetter);
-  return { ...headers, [TRACE_HEADER]: toPayload(defaultPayloadConverter, carrier) };
+  return { ...headers, [TRACE_HEADER]: toPayload(payloadConverter, carrier) };
```
Eventually I think we should just make interceptor headers `Record<string, unknown>` and apply the converter in framework code.
Not 100% sure about that though.

bergundy commented on packages/test/src/test-custom-payload-converter.ts:
```diff
@@ -56,6 +58,122 @@ if (RUN_INTEGRATION_TESTS) {
+    );
+  });
+
+  test('fromPayload throws from client.execute', async (t) => {
```
I'd just add a clarification that throwing is done when getting the result.
Not critical.

bergundy reviewed (approved):
Great stuff!
